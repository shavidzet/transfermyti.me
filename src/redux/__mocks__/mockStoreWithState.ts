import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
