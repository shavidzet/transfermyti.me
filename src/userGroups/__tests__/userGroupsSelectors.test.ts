import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as userGroupsSelectors from "../userGroupsSelectors";

const TEST_STATE = {
  ...state,
  userGroups: {
    source: {
      "5001": {
        id: "5001",
        name: "Test Group A",
        userIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: "clock-user-group-02",
        isIncluded: false,
        memberOf: "userGroups",
      },
      "5002": {
        id: "5002",
        name: "Test Group B",
        userIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: "userGroups",
      },
    },
    target: {
      "clock-user-group-01": {
        id: "clock-user-group-01",
        name: "Admins",
        userIds: ["clock-user-01"],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: "userGroups",
      },
      "clock-user-group-02": {
        id: "clock-user-group-02",
        name: "Test Group A",
        userIds: [],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "5001",
        isIncluded: false,
        memberOf: "userGroups",
      },
    },
    isFetching: false,
  },
};

describe("within userGroupsSelectors", () => {
  test("the sourceUserGroupsSelector returns array of values from state.source", () => {
    const result = userGroupsSelectors.sourceUserGroupsSelector(TEST_STATE);

    expect(result).toEqual(Object.values(TEST_STATE.userGroups.source));
  });

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceUserGroupsSelector",
        selector: userGroupsSelectors.includedSourceUserGroupsSelector,
      },
      {
        name: "for the sourceUserGroupsForTransferSelector",
        selector: userGroupsSelectors.sourceUserGroupsForTransferSelector,
      },
      {
        name: "for the sourceUserGroupsInActiveWorkspaceSelector",
        selector: userGroupsSelectors.sourceUserGroupsInActiveWorkspaceSelector,
      },
    ],
  );
});
