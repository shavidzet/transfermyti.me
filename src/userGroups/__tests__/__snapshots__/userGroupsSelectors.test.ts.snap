// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within userGroupsSelectors the selectors match their snapshots for the includedSourceUserGroupsSelector 1`] = `
Array [
  Object {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": Array [],
    "workspaceId": "1001",
  },
]
`;

exports[`within userGroupsSelectors the selectors match their snapshots for the sourceUserGroupsForTransferSelector 1`] = `
Array [
  Object {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": Array [],
    "workspaceId": "1001",
  },
]
`;

exports[`within userGroupsSelectors the selectors match their snapshots for the sourceUserGroupsInActiveWorkspaceSelector 1`] = `
Array [
  Object {
    "entryCount": 0,
    "id": "5001",
    "isIncluded": false,
    "linkedId": "clock-user-group-02",
    "memberOf": "userGroups",
    "name": "Test Group A",
    "userIds": Array [],
    "workspaceId": "1001",
  },
  Object {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": Array [],
    "workspaceId": "1001",
  },
]
`;
